### ISSUE TITLE
Reindex daily indices to weekly index

### ISSUE THREAD

**Issue by panda87** [elastic/curator#935]
Hi

Thanks for the latest new actions - rollover and reindex!

I see there is an example of reindex that shows how to merge the daily indices, but this is lack the capability of how to tell the "dest" conf file the current week number in an automated way.

Also, once the action is done, can we add a flag that will delete the old merged daily indices?
I'm fully aware the risk, but maybe there is an option to add a basic validation in the end of the action to make sure all the results merged.

Thanks
D.

**Comment by untergeek** [comment]
@panda87 
> but this is lack the capability of how to tell the "dest" conf file the current week number in an automated way.

It would be great if [date math](https://www.elastic.co/guide/en/elasticsearch/reference/current/date-math-index-names.html) were supported for index names in `dest`, but it isn't.  I _can_ make it possible to do strftime renaming, like with the [name option](https://www.elastic.co/guide/en/elasticsearch/client/curator/current/option_name.html) for the create_index action, but it won't do any date-math.  

Would that be acceptable?  It would only allow for the current week to be shown, so it would not necessarily look correct, as "this" week would contain "last" week's data.

> Also, once the action is done, can we add a flag that will delete the old merged daily indices?

This is as simple as just making another action with the same filters, but with `delete_indices` being the action.  I would ensure that the `reindex` action has `continue_if_exception: False`, so that it can't proceed if there's an error.

> maybe there is an option to add a basic validation in the end of the action to make sure all the results merged.

The reindex action already [checks to make sure the `dest` index is present](https://github.com/elastic/curator/blob/v5.0.1/curator/actions.py#L1139-L1158), but there is no conclusive way, in or out of elasticsearch, to _validate_ the reindex.  This is mostly because a reindex might involve a query, and only reindex a subset of results. Or it might be merging all documents from multiple.  Or you might be reformatting the data through an ingest node. If you've specified one of these, how can Curator know what the resulting index should look like?

A safer bet would be a "temporary" snapshot and perhaps a dedicated repository, too.  That way you can snapshot that week's worth of data, and then do some other validation on the merged index.  Delete the temporary snapshot when you're satisfied.

**Comment by joskfg** [comment]
I did something like you comment. I moved information from the last day daily index to monthly index using a watcher like this:
```
PUT _xpack/watcher/watch/my_watcher
{
  "trigger": {
    "schedule": {
      "daily" : { "at" : "01:00" }
    }
  },
  "input": {
    "search": {
      "request": {
        "indices": [
          "<my_index-{now-1d\/d}>"
        ],
        "body": {
          "query": {
            "match_all": {}
          }
        }
      }
    }
  },
  "actions": {
    "index_payload": {
      "index": {
        "index": "<my_index-{now-1d\/M{YYYY.MM}}>",
        "doc_type": "my_type"
      }
    }
  }
}
```

This watcher move the data from `<my_index-{now-1d\/d}> (my_index-2017.04.20)` to `<my_index-{now-1d\/d}> (my_index-2017.04)`. 

With this you have the reindex part (you can add transformation too), but you cannot remove the index from here, so I used the [curator_cli](https://www.elastic.co/guide/en/elasticsearch/client/curator/current/singleton-cli.html) to remove it using a cron that runs every day with the curator execition 
```
curator_cli --use_ssl --dry-run --host="https://my-elasticsearch-host" delete_indices --ignore_empty_list --filter_list '[{"filtertype":"age","source":"creation_date","direction":"older","unit":"days","unit_count":2},{"filtertype":"pattern", "kind":"regex", "value":"^my_index-[a-z]{2}-\\d{4}.\\d{2}.\\d{2}$$"}]'`.
```

This will remove all indices older than 2 days that have that pattern. For example two days after the creation of `my_index-2017.04.20`, this index will be remove. I let it two days just for safety.

I hope this help you.

**Comment by untergeek** [comment]
@joskfg This is great for people using Watcher, but that's a commercial product and not everyone has access to it.

**Comment by untergeek** [comment]
> It would be great if date math were supported for index names in dest, but it isn't. I can make it possible to do strftime renaming, like with the name option for the create_index action, but it won't do any date-math.

Apparently, date math _is_ supported in `dest`.  This was just addressed properly in #1020

**Comment by masahirotoriumi** [comment]
+1

**Comment by temporafugiunt** [comment]
Little late to the party... But...

We needed to consolidate daily indexes into monthly indexes in an easy command line based way, and we are a .NET shop, so we created a command line application and a nuget package for the client library DLL for this purpose. The source is available here and can be compiled in any Visual Studio 2017 version:

https://github.com/trackitsvalue/indexify

We love curator, but it was actually more than what was necessary to perform _**this specific function**_ so we created this specialized tool for this task. Pull requests and additions are always welcome.

Cheers!
